lib/evaluate/tools.h

```diff
@@ -678,32 +683,36 @@ struct TypeKindVisitor {
 // GetLastSymbol() returns the rightmost symbol in an object or procedure
 // designator (which has perhaps been wrapped in an Expr<>), or a null pointer
 // when none is found.
-struct GetLastSymbolVisitor
-  : public virtual VisitorBase<std::optional<const semantics::Symbol *>> {
-  // std::optional<> is used because it is default-constructible.
+struct GetLastSymbolHelper : public AnyTraverse<GetLastSymbolHelper,
+                                 std::optional<const semantics::Symbol *>> {
   using Result = std::optional<const semantics::Symbol *>;
-  explicit GetLastSymbolVisitor(std::nullptr_t) {}
-  void Handle(const semantics::Symbol &x) { Return(&x); }
-  void Handle(const Component &x) { Return(&x.GetLastSymbol()); }
-  void Handle(const NamedEntity &x) { Return(&x.GetLastSymbol()); }
-  void Handle(const ProcedureDesignator &x) { Return(x.GetSymbol()); }
-  template<TypeCategory CAT, int KIND>
-  void Pre(const Expr<Type<CAT, KIND>> &x) {
-    if (!std::holds_alternative<Designator<Type<CAT, KIND>>>(x.u)) {
-      Return(nullptr);
-    }
+  using Base = AnyTraverse<GetLastSymbolHelper, Result>;
+  GetLastSymbolHelper() : Base{*this} {}
+  using Base::operator();
+  Result operator()(const semantics::Symbol &x) const { return &x; }
+  Result operator()(const Component &x) const { return &x.GetLastSymbol(); }
+  Result operator()(const NamedEntity &x) const { return &x.GetLastSymbol(); }
+  Result operator()(const ProcedureDesignator &x) const {
+    return x.GetSymbol();
   }
-  void Pre(const Expr<SomeDerived> &x) {
-    if (!std::holds_alternative<Designator<SomeDerived>>(x.u)) {
-      Return(nullptr);
+  template<typename T> Result operator()(const Expr<T> &x) const {
+    if constexpr (common::HasMember<T, AllIntrinsicTypes> ||
+        std::is_same_v<T, SomeDerived>) {
+      if (const auto *designator{std::get_if<Designator<T>>(&x.u)}) {
+        if (auto known{(*this)(*designator)}) {
+          return known;
+        }
+      }
+      return nullptr;
```

Comment:
Shouldn't this be `return {};` ? It looks weird to me to have optional containing `nullptr`, though I agree it should be functionally equivalent given how `GetLastSymbolHelper ` is currently used.
I think you could also directly have `return (*this)(*designator)` above.

Reply:
`return {};` would be the same as `return std::nullopt;`, meaning that the result of the search remains unknown.  `return nullptr;` means that the result of the search is known and that it has failed.